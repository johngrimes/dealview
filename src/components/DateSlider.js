import React from 'react'
import PropTypes from 'prop-types'
import ReactSlider from 'react-slider'
import throttle from 'lodash.throttle'

import './styles/DateSlider.css'

class DateSlider extends React.Component {
  static propTypes = {
    dates: PropTypes.arrayOf(PropTypes.string).isRequired,
    selected: PropTypes.string,
    initialised: PropTypes.bool.isRequired,
    className: PropTypes.string,
    onChange: PropTypes.func,
  }

  constructor(props) {
    super(props)

    const selectedIndex = props.dates.indexOf(props.selected)
    this.state = {
      selected: props.selected,
      selectedIndex: selectedIndex !== -1 ? selectedIndex : 0,
    }

    this.handleChange = this.handleChange.bind(this)
  }

  handleChangeUnthrottled(value) {
    const selectedIndex = parseInt(value, 10)
    const selected = this.props.dates[selectedIndex]
    this.setState(
      () => ({ selectedIndex, selected }),
      () => {
        if (this.props.onChange) this.props.onChange(selected)
      }
    )
  }
  handleChange = throttle(this.handleChangeUnthrottled, 100)

  componentWillReceiveProps(props) {
    const selectedIndex = props.dates.indexOf(props.selected)
    this.setState(() => ({
      selected: props.selected,
      selectedIndex: selectedIndex !== -1 ? selectedIndex : 0,
    }))
  }

  render() {
    const { dates, className, initialised } = this.props
    const { selected, selectedIndex } = this.state
    return (
      <div
        className={
          initialised ? 'date-slider' : 'date-slider date-slider-uninitialised'
        }
      >
        <div className='date-tag start-date'>
          {dates[0]}
        </div>
        <div className='date-tag end-date'>
          {dates[dates.length - 1]}
        </div>
        <ReactSlider
          className={className ? `slider ${className}` : className}
          min={0}
          max={dates.length - 1}
          defaultValue={0}
          value={selectedIndex}
          onChange={this.handleChange}
        >
          <div className='date-tag selected-date'>
            {selected}
          </div>
        </ReactSlider>
      </div>
    )
  }
}

export default DateSlider
