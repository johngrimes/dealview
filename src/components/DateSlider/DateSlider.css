.slider {
  width: 500px;
  margin: 20px 0 20px;
  display: block;
}
