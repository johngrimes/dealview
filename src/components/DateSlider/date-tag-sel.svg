<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 60.01 17.87"><defs><style>.cls-1{fill:#d4cbca;}</style></defs><title>date-tag-sel_1</title><g id="Layer_2" data-name="Layer 2"><g id="Layer_1-2" data-name="Layer 1"><path class="cls-1" d="M3,4H26l4-4,4,4H57a3,3,0,0,1,3,3v7.84a3,3,0,0,1-3,3H3a3,3,0,0,1-3-3V7A3,3,0,0,1,3,4Z"/></g></g></svg>