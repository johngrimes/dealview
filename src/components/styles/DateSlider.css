.slider {
  clear: both;
  margin: 34px 0 40px;
  height: 8px;
  background: linear-gradient(#aaa, #ddd);
  display: block;
  border-radius: 6px;
}

.handle {
  background: linear-gradient(#129fea, #216ba5);
  width: 20px;
  height: 20px;
  border-radius: 10px;
  box-sizing: border-box;
  box-shadow: 0px 1px 5px 1px #aaa;
  position: relative;
  top: -6px;
}

.date-tag {
  background: transparent url('../images/date-tag.svg') no-repeat top left;
  width: 76px;
  background-size: 76px;
  font-size: 13px;
  text-align: center;
  height: 30px;
}

.start-date, .end-date {
  margin-top: 10px;
  margin-bottom: -6px;
  position: relative;
}

.start-date {
  float: left;
  left: -36px;
}

.end-date {
  float: right;
  left: 36px;
}

.selected-date {
  background: transparent url('../images/date-tag-sel.svg') no-repeat top left;
  line-height: 28px;
  position: relative;
  left: -28px;
  top: 21px;
}

.date-slider-uninitialised .date-tag,
.date-slider-uninitialised .handle {
  display: none;
}
