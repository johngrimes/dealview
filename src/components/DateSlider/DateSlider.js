// @flow

import React from 'react'
import ReactSlider from 'react-slider'

import 'components/DateSlider/DateSlider.css'

type Props = {
  +dates: string[],
  +selected: string,
  +initialised: boolean,
  +className?: string,
  +onChange?: (value: string) => void,
}
type State = {
  +selected: string,
  +selectedIndex: number,
}

class DateSlider extends React.Component {
  props: Props
  state: State
  handleChange: (value: string) => void

  constructor(props: Props) {
    super(props)

    const selectedIndex = props.dates.indexOf(props.selected)
    this.state = {
      selected: props.selected,
      selectedIndex: selectedIndex !== -1 ? selectedIndex : 0,
    }

    this.handleChange = this.handleChange.bind(this)
  }

  handleChange(value: string): void {
    const selectedIndex = parseInt(value)
    const selected = this.props.dates[selectedIndex]
    this.setState(
      () => ({ selectedIndex, selected }),
      () => { if (this.props.onChange) this.props.onChange(selected) }
    )
  }

  componentWillReceiveProps(props: Props) {
    const selectedIndex = props.dates.indexOf(props.selected)
    this.setState(() => ({
      selected: props.selected,
      selectedIndex: selectedIndex !== -1 ? selectedIndex : 0,
    }))
  }

  render() {
    const { dates, className, initialised } = this.props
    const { selected, selectedIndex } = this.state
    return (
      <div className={initialised ? 'date-slider' : 'date-slider date-slider-uninitialised'}>
        <div className='date-tag start-date'>{dates[0]}</div>
        <div className='date-tag end-date'>{dates[dates.length - 1]}</div>
        <ReactSlider className={className ? `slider ${className}` : className}
          min={0} max={dates.length - 1} defaultValue={0}
          value={selectedIndex}
          onChange={this.handleChange}>
          <div className='date-tag selected-date'>{selected}</div>
        </ReactSlider>
      </div>
    )
  }
}

export default DateSlider
