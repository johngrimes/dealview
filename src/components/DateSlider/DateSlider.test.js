/* global expect */

import React from 'react'
import { shallow } from 'enzyme'

import DateSlider from 'components/DateSlider/DateSlider'

describe('DateSlider', () => {
  it('should render correctly', () => {
    const props = {
      dates: [ '2017-06-21', '2017-06-22', '2017-06-23' ],
      selected: '2017-06-22',
      initialised: true,
      className: 'someClass',
    }
    const wrapper = shallow(<DateSlider {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})
