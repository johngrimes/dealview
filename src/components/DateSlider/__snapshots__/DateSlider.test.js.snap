// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateSlider should render correctly 1`] = `
<div
  className="date-slider"
>
  <div
    className="date-tag start-date"
  >
    2017-06-21
  </div>
  <div
    className="date-tag end-date"
  >
    2017-06-23
  </div>
  <ReactSlider
    barClassName="bar"
    className="slider someClass"
    defaultValue={0}
    disabled={false}
    handleActiveClassName="active"
    handleClassName="handle"
    invert={false}
    max={2}
    min={0}
    minDistance={0}
    onChange={[Function]}
    orientation="horizontal"
    pearling={false}
    snapDragDisabled={false}
    step={1}
    value={1}
    withBars={false}
  >
    <div
      className="date-tag selected-date"
    >
      2017-06-22
    </div>
  </ReactSlider>
</div>
`;
